modules/api.py

```diff
@@ -0,0 +1,170 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+from __future__ import annotations
+
+import asyncio
+import logging
+from typing import Any
+
+import aiohttp
+from discord.backoff import ExponentialBackoff
+
+import core
+from constants import (
+    PAPIWebsocketCloseCodes,
+    PAPIWebsocketNotificationTypes,
+    PAPIWebsocketOPCodes,
+    PAPIWebsocketSubscriptions,
+)
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+WS_URL: str = "wss://api.pythonista.gg/v1/websocket"
+
+
+class API(core.Cog):
+    def __init__(self, bot: core.Bot) -> None:
+        self.bot = bot
+
+        self.backoff: ExponentialBackoff = ExponentialBackoff()  # type: ignore
+        self.websocket: aiohttp.ClientWebSocketResponse | None = None
+
+        self.connection_task: asyncio.Task[None] | None = None
+        self.keep_alive_task: asyncio.Task[None] | None = None
+
+    async def cog_load(self) -> None:
+        self.connection_task = asyncio.create_task(self.connect())
+
+    async def cog_unload(self) -> None:
+        if self.connection_task:
+            try:
+                self.connection_task.cancel()
+            except Exception as e:
+                LOGGER.debug(f'Unable to cancel Pythonista API connection_task in "cog_unload": {e}')
+
+        if self.is_connected():
+            assert self.websocket
+            await self.websocket.close(code=PAPIWebsocketCloseCodes.NORMAL)
+
+        if self.keep_alive_task:
+            try:
+                self.keep_alive_task.cancel()
+            except Exception as e:
+                LOGGER.debug(f'Unable to cancel Pythonista API keep_alive_task in "cog_unload": {e}')
+
+    def is_connected(self) -> bool:
+        return self.websocket is not None and not self.websocket.closed
+
+    async def connect(self) -> None:
+        token: str | None = core.CONFIG["TOKENS"].get("pythonista")
+
+        if not token:
+            self.connection_task = None
+            return
+
+        if self.keep_alive_task:
+            try:
+                self.keep_alive_task.cancel()
+            except Exception as e:
+                LOGGER.debug(f"Failed to cancel Pythonista API Websocket keep alive. This is likely not a problem: {e}")
+
+        headers: dict[str, str] = {"Authorization": token}
+
+        while True:
+            async with aiohttp.ClientSession(headers=headers) as session:
```

Comment:
Do we need to make the session on every iteration of this `while`?

Reply:
No, good catch.